File: ios/OTAManager.swift
import Foundation

public class OTAManager {

    private let bleClient: BleClient

    public init(bleClient: BleClient) {
        self.bleClient = bleClient
    }

    public func checkFirmwareVersion(device: Device) async -> Result<String, PlaudError> {
        do {
            // Command 0x09: Get software version
            let command = Data([0x5A, 0x03, 0x09])
            bleClient.sendCommand(command: command)
            // Response parsing logic will be implemented
            return .success("1.0.0")
        } catch {
            return .failure(PlaudError(code: 6001, message: "Check firmware version failed: \(error.localizedDescription)"))
        }
    }

    public func startOTAUpdate(device: Device, firmwareFile: URL, progressCallback: @escaping (Int) -> Void = { _ in }) async -> Result<Void, PlaudError> {
        do {
            // OTA upgrade process:
            // 1. Query firmware version (0x88)
            // 2. Send firmware package (0x89) - may be called multiple times for large files
            // 3. Start OTA upgrade (0x90)
            // 4. Device reboots

            // Step 1: Query version
            let queryCommand = Data([0x5A, 0x03, 0x88])
            bleClient.sendCommand(command: queryCommand)

            // Step 2: Send firmware package
            let fileData = try Data(contentsOf: firmwareFile)
            let chunkSize = 240 // Max payload size

            var offset = 0
            while offset < fileData.count {
                let chunkEnd = min(offset + chunkSize, fileData.count)
                let chunk = fileData.subdata(in: offset..<chunkEnd)

                var sendCommand = Data([0x5A, UInt8(3 + chunk.count), 0x89])
                sendCommand.append(chunk)

                bleClient.sendCommand(command: sendCommand)
                offset = chunkEnd
                progressCallback((offset * 100) / fileData.count)
            }

            // Step 3: Start OTA upgrade
            let upgradeCommand = Data([0x5A, 0x03, 0x90])
            bleClient.sendCommand(command: upgradeCommand)

            progressCallback(100)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 6002, message: "OTA update failed: \(error.localizedDescription)"))
        }
    }

    public func syncDeviceTime(device: Device) async -> Result<Void, PlaudError> {
        do {
            // Command 0x91: Sync RTC time
            let currentTime = UInt32(Date().timeIntervalSince1970)
            let command = Data([
                0x5A,
                0x07,
                0x91,
                UInt8((currentTime >> 24) & 0xFF),
                UInt8((currentTime >> 16) & 0xFF),
                UInt8((currentTime >> 8) & 0xFF),
                UInt8(currentTime & 0xFF)
            ])

            bleClient.sendCommand(command: command)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 6003, message: "Sync device time failed: \(error.localizedDescription)"))
        }
    }
}

File: android/src/main/java/com/memotalk/sdk/OTAManager.kt
package com.memotalk.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

class OTAManager(private val bleClient: BleClient) {

    suspend fun checkFirmwareVersion(device: Device): Result<String, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x09: Get software version
            val command = ByteArray(3)
            command[0] = 0x5A // Frame header
            command[1] = 0x03 // Length
            command[2] = 0x09 // Command

            bleClient.sendCommand(command)
            // Response parsing logic will be implemented
            Result.success("1.0.0")
        } catch (e: Exception) {
            Result.failure(PlaudError(6001, "Check firmware version failed: ${e.message}"))
        }
    }

    suspend fun startOTAUpdate(device: Device, firmwareFile: File, progressCallback: (progress: Int) -> Unit = {}): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // OTA upgrade process:
            // 1. Query firmware version (0x88)
            // 2. Send firmware package (0x89) - may be called multiple times for large files
            // 3. Start OTA upgrade (0x90)
            // 4. Device reboots

            // Step 1: Query version
            val queryCommand = ByteArray(3)
            queryCommand[0] = 0x5A // Frame header
            queryCommand[1] = 0x03 // Length
            queryCommand[2] = 0x88 // Command
            bleClient.sendCommand(queryCommand)

            // Step 2: Send firmware package
            val fileSize = firmwareFile.length()
            val fileData = firmwareFile.readBytes()

            // Split into chunks if necessary (MTU limit)
            val chunkSize = 240 // Max payload size
            var offset = 0
            while (offset < fileData.size) {
                val chunkEnd = minOf(offset + chunkSize, fileData.size)
                val chunk = fileData.sliceArray(offset until chunkEnd)

                val sendCommand = ByteArray(3 + chunk.size)
                sendCommand[0] = 0x5A // Frame header
                sendCommand[1] = (3 + chunk.size).toByte() // Length
                sendCommand[2] = 0x89 // Command
                chunk.forEachIndexed { index, byte -> sendCommand[3 + index] = byte }

                bleClient.sendCommand(sendCommand)
                offset = chunkEnd
                progressCallback((offset * 100) / fileData.size)
            }

            // Step 3: Start OTA upgrade
            val upgradeCommand = ByteArray(3)
            upgradeCommand[0] = 0x5A // Frame header
            upgradeCommand[1] = 0x03 // Length
            upgradeCommand[2] = 0x90 // Command
            bleClient.sendCommand(upgradeCommand)

            progressCallback(100)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(6002, "OTA update failed: ${e.message}"))
        }
    }

    suspend fun syncDeviceTime(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Command 0x91: Sync RTC time
            val currentTime = System.currentTimeMillis() / 1000 // Unix timestamp
            val command = ByteArray(7)
            command[0] = 0x5A // Frame header
            command[1] = 0x07 // Length
            command[2] = 0x91 // Command
            // Add 4-byte timestamp
            command[3] = (currentTime shr 24).toByte()
            command[4] = (currentTime shr 16).toByte()
            command[5] = (currentTime shr 8).toByte()
            command[6] = currentTime.toByte()

            bleClient.sendCommand(command)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(6003, "Sync device time failed: ${e.message}"))
        }
    }
}
